File: leetcode/editor/en/[931]Minimum Falling Path Sum.java
931
        Minimum Falling Path Sum
        2022-12-13 09:41:58

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        int[][] dp = new int[n][n];
        for (int i = 0; i < n; i++) {
            dp[i] = matrix[i].clone();
        }
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (j == 0) {
                    dp[i][j] += Math.min(dp[i - 1][j], dp[i - 1][j + 1]);
                } else if (j == n - 1) {
                    dp[i][j] += Math.min(dp[i - 1][j], dp[i - 1][j - 1]);
                } else {
                    dp[i][j] += Math.min(dp[i - 1][j], Math.min(dp[i - 1][j - 1], dp[i - 1][j + 1]));
                }
            }
        }
        int res = Integer.MAX_VALUE;
        for (int num : dp[n - 1]) {
            res = Math.min(res, num);
        }
        return res;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[931]Minimum Falling Path Sum.js
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * @param {number[][]} matrix
 * @return {number}
 */
var minFallingPathSum = function (matrix) {
    let n = matrix.length;
    for (let row = 1; row < n; row++) {
        for (let col = 0; col < n; col++) {
            let min = matrix[row - 1][col];

            if (col > 0) {
                min = Math.min(min, matrix[row - 1][col - 1]);
            }

            if (col < n - 1) {
                min = Math.min(min, matrix[row - 1][col + 1]);
            }

            matrix[row][col] += min;
        }
    }

    return Math.min(...matrix[n - 1]);
};
//leetcode submit region end(Prohibit modification and deletion)
